backend/src/db/migrations/20200125010142-refactor_all_images_to_separate_type.js

```diff
@@ -0,0 +1,101 @@
+/* eslint-disable no-console */
+import { getDriver } from '../../db/neo4j'
+
+export const description = `
+  Refactor all our image properties on posts and users to a dedicated type
+  "Image" which contains metadata and image file urls.
+`
+
+const printSummaries = summaries => {
+  console.log('=========================================')
+  summaries.forEach(stat => {
+    console.log(stat.query.text)
+    console.log(JSON.stringify(stat.counters, null, 2))
+  })
+  console.log('=========================================')
+}
+
+export async function up() {
+  const driver = getDriver()
+  const session = driver.session()
+  const writeTxResultPromise = session.writeTransaction(async txc => {
+    const runs = await Promise.all(
+      [
+        `
+      MATCH (post:Post)
+      WHERE post.image IS NOT NULL
+      CREATE (post)-[:HERO_IMAGE]->(image:Image)
+      SET
+        image.url         = post.image,
+        image.sensitive   = post.imageBlurred,
+        image.aspectRatio = post.imageAspectRatio
+      REMOVE
+        post.image,
+        post.imageBlurred,
+        post.imageAspectRatio
+    `,
+        `
+      MATCH (user:User)
+      WHERE user.avatar IS NOT NULL
+      CREATE (user)-[:AVATAR_IMAGE]->(avatar:Image)
+      SET avatar.url = user.avatar
+      REMOVE user.avatar
+    `,
+        `
+      MATCH (user:User)
+      WHERE user.coverImg IS NOT NULL
+      CREATE (user)-[:COVER_IMAGE]->(coverImage:Image)
+      SET coverImage.url = user.coverImg
+      REMOVE user.coverImg
+    `,
+      ].map(s => txc.run(s)),
+    )
+    return runs.map(({ summary }) => summary)
+  })
+
+  try {
+    const stats = await writeTxResultPromise
+    console.log('Created image nodes from all user avatars and post images.')
+    printSummaries(stats)
+  } finally {
+    session.close()
+  }
+}
+
+export async function down() {
+  const driver = getDriver()
+  const session = driver.session()
+  const writeTxResultPromise = session.writeTransaction(async txc => {
+    const runs = await Promise.all(
+      [
+        `
+      MATCH (post)-[:HERO_IMAGE]->(image:Image)
+      SET
+        post.image            = image.url,
+        post.imageBlurred     = image.sensitive,
+        post.imageAspectRatio = image.aspectRatio
+      DETACH DELETE image
+    `,
+        `
+      MATCH(user)-[:AVATAR_IMAGE]->(avatar:Image)
+      SET user.avatar = avatar.url
+      DETACH DELETE avatar
+    `,
+        `
+      MATCH(user)-[:COVER_IMAGE]->(coverImage:Image)
+      SET user.coverImg = coverImage.url
+      DETACH DELETE coverImage
```

Comment:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2020-02-11T23:56:13Z" title="Wednesday, February 12th 2020, 12:56:13 am +01:00">Feb 12, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,101 @@
+/* eslint-disable no-console */
+import { getDriver } from '../../db/neo4j'
+
+export const description = `
+  Refactor all our image properties on posts and users to a dedicated type
+  "Image" which contains metadata and image file urls.
+`
+
+const printSummaries = summaries => {
+  console.log('=========================================')
+  summaries.forEach(stat => {
+    console.log(stat.query.text)
+    console.log(JSON.stringify(stat.counters, null, 2))
+  })
+  console.log('=========================================')
+}
+
+export async function up() {
+  const driver = getDriver()
+  const session = driver.session()
+  const writeTxResultPromise = session.writeTransaction(async txc => {
+    const runs = await Promise.all(
+      [
+        `
+      MATCH (post:Post)
+      WHERE post.image IS NOT NULL
+      CREATE (post)-[:TEASER_IMAGE]->(image:Image)
+      SET
+        image.url         = post.image,
+        image.blurred     = post.imageBlurred,
+        image.aspectRatio = post.imageAspectRatio
+      REMOVE
+        post.image,
+        post.imageBlurred,
+        post.imageAspectRatio
+    `,
+        `
+      MATCH (user:User)
+      WHERE user.avatar IS NOT NULL
+      CREATE (user)-[:AVATAR_IMAGE]->(avatar:Image)
+      SET avatar.url = user.avatar
+      REMOVE user.avatar
+    `,
+        `
+      MATCH (user:User)
+      WHERE user.coverImg IS NOT NULL
+      CREATE (user)-[:COVER_IMAGE]->(coverImage:Image)
+      SET coverImage.url = user.coverImg
+      REMOVE user.coverImg
+    `,
+      ].map(s => txc.run(s)),
+    )
+    return runs.map(({ summary }) => summary)
+  })
+
+  try {
+    const stats = await writeTxResultPromise
+    console.log('Splitted all images from users and posts.')
+    printSummaries(stats)
+  } finally {
+    session.close()
+  }
+}
+
+export async function down() {
+  const driver = getDriver()
+  const session = driver.session()
+  const writeTxResultPromise = session.writeTransaction(async txc => {
+    const runs = await Promise.all(
+      [
+        `
+      MATCH (post)-[:TEASER_IMAGE]->(image:Image)
+      SET
+        post.image            = image.url,
+        post.imageBlurred     = image.blurred,
+        post.imageAspectRatio = image.aspectRatio
+      DETACH DELETE image
+    `,
+        `
+      MATCH(user)-[:AVATAR_IMAGE]->(avatar:Image)
+      SET user.avatar = avatar.url
+      DETACH DELETE avatar
+    `,
+        `
+      MATCH(user)-[:COVER_IMAGE]->(coverImage:Image)
+      SET user.coverImg = coverImage.url
+      DETACH DELETE coverImage
```


Maybe just `REMOVE user.coverImg`.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-02-20T19:36:57Z" title="Thursday, February 20th 2020, 8:36:57 pm +01:00">Feb 20, 2020</time>_

---


good question... we can discuss